File: C++/Container With Most Water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        if(height.size() < 2)
            return 0;
        
        // greedy method
        int left = 0, right = height.size() - 1;
        int max_vol = 0, current_vol;
        while(left < right){
            // always determined by the shorest side
            current_vol = min(height[left], height[right]) * (right - left);
            
            max_vol = max(max_vol, current_vol);
            

            // does not matter if left == right, cuz either part will make no contribution later on
            if(height[left] < height[right])
                left += 1;
            else
                right -= 1;
        }
        
        return max_vol;
    }
};

File: Python/Container With Most Water.py
class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        
        # O(N) -> greedy method
        left, right = 0, len(height) - 1
        max_area = 0
        while left < right:
            min_element = min(height[left], height[right])
            max_area = max(max_area, min_element * (right - left))
            
            if min_element == height[left]:
                left += 1
            else:
                right -= 1
        
        return max_area
        
        """
        # O(n^2) with pruning
        max_area = 0
        length = len(height)
        for index in range(length):
            max_limit = height[index] * (length - index - 1)
            if max_area > max_limit:
                continue
            
            for other_index in range(1, length - index):
                width = length - index - other_index
                if max_area > height[index] * width:
                    break
                
                area = min(height[index], height[-other_index]) * width
                if area > max_area:
                    max_area = area
        
        return max_area
        """